# MLB Batters Multiple Linear Regression - Statsmodels

In [1]:
# import required modules
import pandas as pd
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix
import pandas as pd
from sklearn.metrics import roc_auc_score
from sklearn.metrics import precision_recall_fscore_support
import numpy as np
import statsmodels.api as sm

In [2]:
# load dataset
data = pd.read_csv('../../Capstone/Data/MLB_Bpipeline1.csv', header=0)
data.shape

(35504, 44)

In [3]:
data.columns

Index(['name_common', 'year_ID', 'team_ID', 'lg_ID', 'G', 'Inn', 'runs_dp',
       'runs_field', 'runs_infield', 'runs_defense', 'runs_position',
       'runs_position_p', 'teamRpG', 'oppRpG', 'oppRpPA_rep', 'oppRpG_rep',
       'pyth_exponent', 'pyth_exponent_rep', 'waa_win_perc_off',
       'waa_win_perc_rep', 'OPS_plus', 'age', 'PA', 'runs_bat', 'runs_br',
       'runs_outfield', 'runs_catcher', 'runs_good_plays', 'runs_replacement',
       'runs_above_rep', 'runs_above_avg', 'runs_above_avg_off',
       'runs_above_avg_def', 'WAA', 'WAA_off', 'WAA_def', 'WAR', 'WAR_def',
       'WAR_off', 'WAR_rep', 'waa_win_perc', 'waa_win_perc_def', 'TOB_lg',
       'TB_lg'],
      dtype='object')

In [4]:
features=['runs_defense', 'teamRpG', 'oppRpG', 'oppRpG_rep', 'pyth_exponent', 'pyth_exponent_rep', 'waa_win_perc_rep',
'runs_above_avg_def', 'WAA_def', 'WAR_def']
X=data[features]
y=data.WAR

In [5]:
# split X and y into training and testing sets
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=2019)

In [6]:
model = sm.OLS(y, X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                    WAR   R-squared (uncentered):                   0.680
Model:                            OLS   Adj. R-squared (uncentered):              0.680
Method:                 Least Squares   F-statistic:                              8393.
Date:                Mon, 22 Jun 2020   Prob (F-statistic):                        0.00
Time:                        02:03:51   Log-Likelihood:                         -29183.
No. Observations:               35504   AIC:                                  5.838e+04
Df Residuals:                   35495   BIC:                                  5.846e+04
Df Model:                           9                                                  
Covariance Type:            nonrobust                                                  
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
runs_defense           0.4321      0.006     75.320      0.000       0.421       0.443
teamRpG                1.9680      0.091     21.605      0.000       1.789       2.147
oppRpG                -3.7795      0.510     -7.417      0.000      -4.778      -2.781
oppRpG_rep             1.8472      0.480      3.849      0.000       0.907       2.788
pyth_exponent         -0.9952      0.173     -5.741      0.000      -1.335      -0.655
pyth_exponent_rep      1.1264      0.178      6.317      0.000       0.777       1.476
waa_win_perc_rep      -0.5304      0.048    -11.017      0.000      -0.625      -0.436
runs_above_avg_def    -0.2976      0.033     -9.128      0.000      -0.361      -0.234
WAA_def                0.1293      0.016      8.013      0.000       0.098       0.161
WAR_def                0.1293      0.016      8.013      0.000       0.098       0.161
==============================================================================
Omnibus:                     6214.410   Durbin-Watson:                   1.722
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            30982.581
Skew:                          -0.764   Prob(JB):                         0.00
Kurtosis:                       7.314   Cond. No.                     1.68e+17
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.14e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

## Equation
WAR = 0.4321(runs_defense) + 1.9680(teamRpG) - 3.7795(oppRpG) + 1.8472(oppRpG_rep) - 0.9952(pyth_exponent) + 1.1264(pyth_exponent_rep) - 0.5304(waa_win_perc_rep) - 0.2976(runs_above_avg_def) + 0.1293(WAA_def) + 0.1293(WAR_def)